llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp

```diff
@@ -1506,6 +1514,8 @@ InstructionCost VPWidenRecipe::computeCost(ElementCount VF,
     return Ctx.TTI.getArithmeticInstrCost(Instruction::Mul, VectorTy,
                                           Ctx.CostKind);
   }
+  case Instruction::ExtractValue:
+    return 0;
```

Comment:
Should this call something like `getVectorInstrCost`?

Would be good to make sure we estimate the cost correctly when the extracts might not be completely free, e.g. if the results are returned via the stack (e.g. something like https://llvm.godbolt.org/z/zsv3G3h7n)

Reply:
I don't think so, I did look around (including `getVectorInstrCost()`), but it does not look like there's a hook for this right now. `getVectorInstrCost()` is currently only for `insertelement` and `extractelement`. I guess I could modify this hook to handle it, or add a new one though? 

https://github.com/llvm/llvm-project/blob/455b3d6df20c9bb50cdba66fd2f3202bc43eb4ac/llvm/lib/Analysis/TargetTransformInfo.cpp#L1079-L1084

Reply:
I've added a new hook `getInsertExtractValueCost` rather than shoehorn it into the vector insert/extract cost functions. By default this just returns 0, but there should be enough information passed to it to implement target-specific costs in the future.

Reply:
Do we have tests to make sure we do the right thing when results are not returned in vector registers (as in  https://llvm.godbolt.org/z/zsv3G3h7n)?

Reply:
If by the "right thing" you mean "Am I costing the return values based on the target-specific call ABI?" then the answer is no. I think the answer to that is _also_ "no" for non-struct-return functions too (I don't think there's any costing based on how the target passes arguments/returns values). 

I don't really think that needs doing in this patch (which just adds basic support). Since nothing is using struct-return vectorization right now. Currently, it needs things to be manually annotated with `vector-function-abi-variant`, and I don't really think your example is a very realistic use case.